dawn/prototype/mylibIconLaplaceDriver.cpp

```diff
@@ -0,0 +1,560 @@
+  // ble) for the cell according to Gauss formula.
+  // It is equal to +1 if the normal to the edge
+  // is outwards from the cell, otherwise is -1.
+  for(const auto& c : mesh.faces()) {
+    int m_sparse = 0;
+    auto [xm, ym] = CellCircumcenter(c);
+    for(const auto& e : c.edges()) {
+      mylib::Vertex vOutside(e->vertex(0).x() - xm, e->vertex(0).y() - ym, -1);
+      edge_orientation_cell(c, m_sparse, level) =
+          sgn(dot(mylib::Vertex(e->vertex(0).x() - xm, e->vertex(0).y() - ym, -1),
+                  mylib::Vertex(primal_normal_x(*e, level), primal_normal_y(*e, level), -1)));
+      m_sparse++;
+      // explanation: the vector cellModpoint -> e.vertex(0) is guaranteed to point outside. The dot
+      // product checks if the edge normal has the same orientation. e.vertex(0) is arbitrary,
+      // vertex(1), or any point on e would work just as well
+    }
+  }
+
+  // init sparse quantities for div and rot
+  for(const auto& v : mesh.vertices()) {
+    int m_sparse = 0;
+    for(const auto& e : v.edges()) {
+      geofac_rot(v, m_sparse, level) = dual_edge_length(*e, level) *
+                                       edge_orientation_vertex(v, m_sparse, level) /
+                                       dual_cell_area(v, level);
+
+      // ptr_int%geofac_rot(jv,je,jb) =                &
+      //    & ptr_patch%edges%dual_edge_length(ile,ibe) * &
+      //    & ptr_patch%verts%edge_orientation(jv,jb,je)/ &
+      //    & ptr_patch%verts%dual_area(jv,jb) * REAL(ifac,wp)
+      m_sparse++;
+    }
+  }
+
+  for(const auto& c : mesh.faces()) {
+    int m_sparse = 0;
+    for(const auto& e : c.edges()) {
+      geofac_div(c, m_sparse, level) = primal_edge_length(*e, level) *
+                                       edge_orientation_cell(c, m_sparse, level) /
+                                       //  ((c.color() == mylib::face_color::downward) ? 1 : -1) /
+                                       cell_area(c, level);
+
+      //  ptr_int%geofac_div(jc,je,jb) = &
+      //    & ptr_patch%edges%primal_edge_length(ile,ibe) * &
+      //    & ptr_patch%cells%edge_orientation(jc,jb,je)  / &
+      //    & ptr_patch%cells%area(jc,jb)
+      m_sparse++;
+    }
+  }
+
+  if(dbg_out) {
+    dumpSparseData(mesh, geofac_rot, level, edgesPerVertex,
+                   std::string("laplICONmylib_geofacRot.txt"));
+    dumpSparseData(mesh, geofac_div, level, edgesPerCell,
+                   std::string("laplICONmylib_geofacDiv.txt"));
+  }
+
+  //===------------------------------------------------------------------------------------------===//
+  // computation starts here
+  //===------------------------------------------------------------------------------------------===//
+
+  dawn_generated::cxxnaiveico::icon<mylibInterface::mylibTag>(
+      mesh, k_size, vec, div_vec, rot_vec, nabla2t1_vec, nabla2t2_vec, nabla2_vec,
+      primal_edge_length, dual_edge_length, tangent_orientation, geofac_rot, geofac_div)
+      .run();
+
```

Comment:
Only difference wrt `myLibIconLaplace.cpp` is here right?

Reply:
right